File: lib/xmldb/org/xmldb/api/tests/CollectionTest.java
package org.xmldb.api.tests;

/*
 *  The XML:DB Initiative Software License, Version 1.0
 *
 *
 * Copyright (c) 2000-2001 The XML:DB Initiative.  All rights
 * reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 *
 * 3. The end-user documentation included with the redistribution,
 *    if any, must include the following acknowledgment:
 *       "This product includes software developed by the
 *        XML:DB Initiative (http://www.xmldb.org/)."
 *    Alternately, this acknowledgment may appear in the software itself,
 *    if and wherever such third-party acknowledgments normally appear.
 *
 * 4. The name "XML:DB Initiative" must not be used to endorse or
 *    promote products derived from this software without prior written
 *    permission. For written permission, please contact info@xmldb.org.
 *
 * 5. Products derived from this software may not be called "XML:DB",
 *    nor may "XML:DB" appear in their name, without prior written
 *    permission of the XML:DB Initiative.
 *
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESSED OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED.  IN NO EVENT SHALL THE APACHE SOFTWARE FOUNDATION OR
 * ITS CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
 * USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
 * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 * ====================================================================
 *
 * This software consists of voluntary contributions made by many
 * individuals on behalf of the XML:DB Initiative. For more information
 * on the XML:DB Initiative, please see <http://www.xmldb.org/>.
 */

import java.io.*;

import junit.framework.*;

import org.xmldb.api.*;
import org.xmldb.api.base.*;
import org.xmldb.api.modules.*;

public class CollectionTest extends XMLDBTestCase {   
   public CollectionTest(String name) {
      super(name);
   }
   
   public static Test suite() {
      return new TestSuite(CollectionTest.class);
   }   
   
   public void testGetName( ) {
      try {             
         assert(col.getName().equals("child1"));
      } catch (Exception e) {
         fail( e.getMessage( ) );
      }
   }
   
   public void testGetParentCollection() {
      try {     
         if ( supportsNested ) {
            Collection child = col.getChildCollection("subchild1");
            
            Collection parent = child.getParentCollection();
                     
            assert(parent.getName().equals("child1"));         
            
            // TODO: need to test behavior on the root collection.
            assert(parent.getParentCollection().getParentCollection() == null);
         }
      } catch (Exception e) {
         fail( e.getMessage( ) );
      }   
   }
   
   public void testGetChildCollectionCount() {
      try {    
         if ( supportsNested ) {         
            assert(col.getChildCollectionCount() == 2);
         }
         else {
            assert(col.getChildCollectionCount() == 0);
         }
      } catch (Exception e) {
         fail( e.getMessage( ) );
      }   
   }
   
   public void testListChildCollections() {
      try {             
         String[] children = col.listChildCollections();
         
         if ( supportsNested ) {
            // We don't want to assume order here
            boolean match1 = false;
            boolean match2 = false;         
            for ( int i = 0; i < children.length; i++ ) {
               String child = children[i];
               if ( child.equals("subchild1") )
                  match1 = true;            
               if ( child.equals("subchild2") )
                  match2 = true;       
            }
            
            assert(match1 && match2);
         
            try {
               assert(children[2] == null);
            }
            catch (ArrayIndexOutOfBoundsException e) {
               // we should pass if we get this exception            
            }
         }
         else {
            assert(children.length == 0);
         }
      } catch (Exception e) {
         fail( e.getMessage( ) );
      }   
   }
   
   
   public void testGetChildCollection() {
      try {             
         if ( supportsNested ) {
            Collection child = col.getChildCollection("subchild1");
            assert(child.getName().equals("subchild1"));
         }
         else {
            assert(col.getChildCollection("subchild1") == null);
         }
      } catch (Exception e) {
         fail( e.getMessage( ) );
      }   
   }
   
   public void testGetResourceCount() {
      try { 
         // Account for the fact the binary resources are optional  
         if ( supportsBinary ) {          
            assert(col.getResourceCount() == 4);
         }
         else {
            assert(col.getResourceCount() == 3);
         }
      } catch (Exception e) {
         fail( e.getMessage( ) );
      }   
   }
   
   public void testListResources() {
      try {             
         String[] resources = col.listResources();
         // We don't want to assume order here
         boolean match1 = false;
         boolean match2 = false;
         boolean match3 = false;
         for ( int i = 0; i < resources.length; i++ ) {
            String res = resources[i];
            if ( res.equals("test1.xml") )
               match1 = true;            
            if ( res.equals("test2.xml") )
               match2 = true;
            if ( res.equals("test3.xml") )
               match3 = true;                  
         }
         
         assert(match1 && match2 && match3);
         
         try {
            // TODO: should account for the fact the binary resources are
            // optional
            assert(resources[4] == null);
         }
         catch (ArrayIndexOutOfBoundsException e) {
            // we should pass if we get this exception            
         }
      } catch (Exception e) {
         fail( e.getMessage( ) );
      }   
   }
   
   public void testCreateResource() {
      try {             
         // Test for XMLResource and explicit id
         XMLResource res = (XMLResource) col.createResource("test4.xml", 
            XMLResource.RESOURCE_TYPE);         
         assert(res.getResourceType().equals(XMLResource.RESOURCE_TYPE));
         assert(res.getId().equals("test4.xml"));
         
         // Test for BinaryResource and auto generate id
         if ( supportsBinary ) {
            BinaryResource res2 = (BinaryResource) col.createResource("", 
               BinaryResource.RESOURCE_TYPE);
            assert(res2.getResourceType().equals(BinaryResource.RESOURCE_TYPE));
            assert(! res2.getId().equals(""));
         }

         // Test for unknown resource type
         try {
            Resource res3 = col.createResource("", "GoogleResource");
            assert(false);
         }
         catch (XMLDBException e) {
            assert(e.errorCode == ErrorCodes.UNKNOWN_RESOURCE_TYPE);
         }
      } catch (Exception e) {
         fail( e.getMessage( ) );
      }   
   }
   
   
   public void testStoreResource() {
      String content = "<?xml version=\"1.0\"?><tag1><tag2>value</tag2></tag1>";
      try {             
         XMLResource res = (XMLResource) col.createResource("", 
            XMLResource.RESOURCE_TYPE);
         res.setContent(content);
         col.storeResource(res);
         
         XMLResource res2 = (XMLResource) col.getResource(res.getId());
         assert(res2 != null);
         assert(res2.getId().equals(res.getId()));
         assert(res2.getContent() != null);
         
         col.removeResource(res);
      } catch (Exception e) {       
         fail( e.getMessage( ) );
      }   
   }
   
   public void testRemoveResource() {
      String content = "<?xml version=\"1.0\"?><tag1><tag2>value</tag2></tag1>"; 
      try {
         // Create a new resource             
         XMLResource res = (XMLResource) col.createResource("", 
            XMLResource.RESOURCE_TYPE);
         res.setContent(content);
         col.storeResource(res);
         
         // Verify the resource exists
         XMLResource res2 = (XMLResource) col.getResource(res.getId());
         assert(res2 != null);
         assert(res2.getId().equals(res.getId()));
         assert(res2.getContent() != null);
         
         // Remove the resource
         col.removeResource(res2);
         Resource res3 = col.getResource(res.getId());
         assert(res3 == null);
         
         // Make sure the resource is gone.
         try {
            col.removeResource(res);
         }
         catch (XMLDBException e) {
            assert(e.errorCode == ErrorCodes.NO_SUCH_RESOURCE);
         }         
      } catch (Exception e) {
         fail( e.getMessage( ) );
      }   
   }
   
   public void testGetResource() {
      try {             
         // Check for an XML resource
         Resource res = col.getResource("test1.xml");
         assert(res != null);
         assert(res.getResourceType().equals(XMLResource.RESOURCE_TYPE));
         assert(res.getContent() != null);
         
         // Check for a binary resource
         if ( supportsBinary ) {
            res = col.getResource("image.gif");
            assert(res != null);
            assert(res.getResourceType().equals(BinaryResource.RESOURCE_TYPE));
            assert(res.getContent() != null);
         }
         
         // Check for a missing resource
         res = col.getResource("missing.xml");
         assert(res == null);         
         
      } catch (Exception e) {
         fail( e.getMessage( ) );
      }   
   }
   
   public void testXPathQueryService() {
      try {             
         if ( supportsXPath ) {
            XPathQueryService serv = 
               (XPathQueryService) col.getService("XPathQueryService", "1.0");
            assert(serv != null);
            
            ResourceSet resultSet = serv.query("/data");
            ResourceIterator result = resultSet.getIterator();
            
            int i = 0;
            while (result.hasMoreResources()) {
               result.nextResource();        
               i++;
            }         
            assert(i == 3);
            
            resultSet = serv.query("/data/child/subchild[@name='subchild1']");
            result = resultSet.getIterator();
            i = 0;
            while (result.hasMoreResources()) {
               result.nextResource();        
               i++;
            }                  
            assert(i == 6);
            
            resultSet = serv.query("/data/child/subchild/subsubchild[@name='subsubchild1']");
            result = resultSet.getIterator();
            i = 0;
            while (result.hasMoreResources()) {
               result.nextResource();        
               i++;
            }                  
            assert(i == 1);
         }
      } catch (Exception e) {         
         fail( e.getMessage( ) );
      }   
   }
   
   public void testStub() {
      try {             
    
      } catch (Exception e) {
         fail( e.getMessage( ) );
      }   
   }
}

File: test/collection_test.rb
require "test/unit"

class CollectionTest < XMLDBTestCase
  def test_get_name()
    assert(col.get_name()=="child1")
  end

  def test_get_parent_collection()
    if ( supports_nested? )
      child = col.get_child_collection("subchild1")
      parent = child.get_parent_collection()
      assert(parent.get_name() == "child1")
      # // TODO: need to test behavior on the root collection.
      assert(parent.get_parentCollection().get_parent_collection().nil?);
    end
  end

  def test_get_child_collection_count()
    if ( supports_nested? )
      assert(col.get_child_collection_count() == 2)
    else
      assert(col.get_child_collection_count() == 0)
    end
  end
  
  def test_list_child_collections() 
    children = col.list_child_collections()

    if ( supports_nested? ) 
      # // We don't want to assume order here
      match1 = false
      match2 = false
      children.each do |child|
        match1 = true if ( child.equals("subchild1") )
        match2 = true if ( child.equals("subchild2") )
      end
      assert(match1 && match2)
      assert(children[2].nil?)
    else 
      assert(children.size == 0);
    end
  end

  def test_get_child_collection() 
    if ( supports_nested? ) 
      child = col.get_child_collection("subchild1")
      assert(child.get_name()=="subchild1")
    else 
      assert(col.get_child_collection("subchild1").nil?);
    end
  end


  def test_get_resource_count()
    # // Account for the fact the binary resources are optional
    if ( supports_binary? )
      assert(col.get_resource_count() == 4)
    else 
      assert(col.get_resource_count() == 3)
    end
  end
  
  def test_list_resources()
    resources = col.list_resources()
    # // We don't want to assume order here
    match1 = false
    match2 = false
    match3 = false
    resources.each do |res|
      res = resources[i];
      match1 = true if ( res=="test1.xml" )
      match2 = true if ( res=="test2.xml" )
      match3 = true if ( res=="test3.xml" )
    end
    assert(match1 && match2 && match3)

    # // TODO: should account for the fact the binary resources are
    # // optional
    assert(resources[4].nil?)
  end
  
  def test_create_resource()
    # // Test for XMLResource and explicit id
    res = col.create_resource("test4.xml",
                              XMLResource::RESOURCE_TYPE)
    assert(res.get_resource_type()==XMLResource::RESOURCE_TYPE)
    assert(res.get_id()=="test4.xml")

    # // Test for BinaryResource and auto generate id
    if ( supports_binary? )
      res2 = col.create_resource("",
                                 BinaryResource::RESOURCE_TYPE)
      assert(res2.get_resource_type()==BinaryResource::RESOURCE_TYPE)
      assert_false(res2.get_id()=="")
    end

    # // Test for unknown resource type
    begin
      res3 = col.create_resource("", "GoogleResource")
      assert(false)
    rescue XMLDBException => e
      assert(e.error_code == ErrorCodes::UNKNOWN_RESOURCE_TYPE)
    end
  end


  def test_store_resource()
    content = "<?xml version=\"1.0\"?><tag1><tag2>value</tag2></tag1>"
    res = col.create_resource("",
                              XMLResource::RESOURCE_TYPE)
    res.set_content(content)
    col.store_resource(res)

    res2 = col.get_resource(res.get_id())
    assert_false(res2.nil?)
    assert(res2.get_id()==res.get_id())
    assert_false(res2.get_content().nil?)

    col.remove_resource(res)
  end
  
  def test_remove_resource()
    content = "<?xml version=\"1.0\"?><tag1><tag2>value</tag2></tag1>"
    res = col.create_resource("",
                              XMLResource::RESOURCE_TYPE)
    res.set_content(content)
    col.store_resource(res)

    res2 = col.get_resource(res.get_id())
    assert_false(res2.nil?)
    assert(res2.get_id()==res.get_id())
    assert_false(res2.get_content().nil?)

    col.remove_resource(res)
    res3 = col.get_resource(res.get_id())
    assert(res3.nil?)

    # // Make sure the resource is gone.
    begin
      col.remove_resource(res)
    rescue XMLDBException=>e
      assert(e.errorCode == ErrorCodes::NO_SUCH_RESOURCE)
    end
  end

  def test_get_resource()
  # // Check for an XML resource
    res = col.get_resource("test1.xml")
    assert_false(res.nil?)
    assert(res.get_resource_type()==XMLResource::RESOURCE_TYPE)
    assert_false(res.get_content().nil?)

  # // Check for a binary resource
    if ( supports_binary? ) 
      res = col.get_resource("image.gif")
      assert_false(res.nil?)
      assert(res.get_resource_type()==BinaryResource::RESOURCE_TYPE)
      assert_false(res.get_content().nil?)
    end

  # // Check for a missing resource
    res = col.get_resource("missing.xml")
    assert(res.nil?)
  end
  
  
  def test_XPath_query_service()
    if ( supports_XPath? ) 
      serv = col.get_service("XPathQueryService", "1.0")
      assert_false(serv.nil?)
      resultSet = serv.query("/data")
      result = resultSet.get_iterator()
      i = 0
      while (result.has_more_resources?) do
        result.next_resource()
        i = i + 1
      end
      assert(i == 3)

      resultSet = serv.query("/data/child/subchild[@name='subchild1']")
      result = resultSet.get_iterator()
      i = 0
      while (result.has_more_resources()) do
        result.next_resource()
        i = i + 1
      end
      assert(i == 6)

      resultSet = serv.query("/data/child/subchild/subsubchild[@name='subsubchild1']")
      result = resultSet.get_iterator()
      i = 0
      while (result.has_more_resources()) do
        result.next_resource()
        i = i + 1
      end
      assert(i == 1)
    end
  end
end